### PR TITLE
caf: leds: Add GPIO LED documentation

### PR DESCRIPTION
**Pull request by MarekPieta**
Change adds information about GPIO based implementation.

Jira: NCSDK-9548

### CODE REVIEW

File: include/caf/leds.rst [hunk 1]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
 
-Additionally, you need to complete the steps described in `Configuring PWM ports and LEDs`_.
+Depending on selected LED driver implementation, you must also enable one of the following sets of options:
+
+* :option:`CONFIG_CAF_LEDS_PWM`, :option:`CONFIG_LED_PWM` and :option:`CONFIG_PWM`
+  These options are required for PWM based implementation.
+  Zephyr's :ref:`zephyr:pwm_api` driver is used for setting the LED color (that is, the brightness of the diodes).
+* :option:`CONFIG_CAF_LEDS_GPIO`, :option:`CONFIG_LED_GPIO` and :option:`CONFIG_GPIO`
+  These options are required for GPIO based implementation.
+  Zephyr's :ref:`zephyr:gpio_api` driver is used for setting the LED color (that is, the brightness of the diodes).
```
Comment (emob-nordic) [inline include/caf/leds.rst:31]: Can we say the brightness here if its only on/off?

File: include/caf/leds.rst [hunk 2]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
 
 Configuration
 *************
 
 To use the module, you must enable the following Kconfig options:
 
 * :option:`CONFIG_CAF_LEDS` - This option enables the LEDs module.
-* :option:`CONFIG_PWM` - This option enables Zephyr's :ref:`zephyr:pwm_api` driver, which is used for setting the LED color (that is, the brightness of the diodes).
 * :option:`CONFIG_LED` - This option enables the LED driver API.
-* :option:`CONFIG_LED_PWM` - This option enabled the LED PWM driver.
```
Comment (emob-nordic) [inline include/caf/leds.rst:24]: Configuration is complex topic right now. I think we should split it into points similarly as in sensor_sampler. Do you agree @greg-fer ?

``` 
To use the module, you must complete the following requirements:

1. Enable the module by following Kconfig options:

* :option:`CONFIG_CAF_LEDS` - This option enables the LEDs module.
* :option:`CONFIG_LED` - This option enables the LED driver API.

#. Chose and configure the driver implementation for controlling LEDs.
   a. PWM - ...
   * :option:`CONFIG_CAF_LEDS_PWM` - ...
   * :option:`CONFIG_LED_PWM` - ....
   * :option:`CONFIG_PWM` - ....
   b. GPIO - ...
   * :option:`CONFIG_CAF_LEDS_GPIO` - ...
   * :option:`CONFIG_LED_GPIO` - ...
   * :option:`CONFIG_GPIO` - ...

```

File: include/caf/leds.rst [hunk 3]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
+Configuring LEDs in DTS
+=======================
+
+You can configure LEDs using compatible DTS nodes.
+The configuration process requires enabling or creating DTS nodes.
+This can be done in the devicetree file, either in the board-specific :file:`dts` file or in a DTS overlay file.
+Using the option with the overlay file will merge the settings to configuration.
+For more help, see :ref:`zephyr:dt-guide`.
+
+The ``compatible`` property value of the DTS node depends on selected LED driver implementation.
```
Comment (greg-fer) [inline include/caf/leds.rst:54]:
```suggestion
The value of the ``compatible`` property of the DTS node depends on the selected LED driver implementation.
```

File: include/caf/leds.rst [hunk 4]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
+For example you can use node that is compatible with ``pwm-leds`` to configure LEDs controlled with PWM.
```
Comment (greg-fer) [inline include/caf/leds.rst:55]:
```suggestion
For example, you can use a node that is compatible with ``pwm-leds`` to configure LEDs controlled with PWM.
```

File: include/caf/leds.rst [hunk 5]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
+See the following subsections for more detailed information about configuring LED driver implementations supported by CAF LEDs.
+
+The LEDs module assumes that a single DTS node is a separate and complete logical LED.
+The LEDs module expects that the node will hold configuration of HW LEDs responsible for reproducing all required color channels.
+The number of HW LEDs configured to reproduce color channels can be either one or three (either monochromatic or following the RGB order, with the red channel defined first, then the green one, then the blue one).
+If only one HW LED is used for a monochromatic setting, the module will convert the tri-channel color to a single value of brightness and pass it to this single HW LED.
```
Comment (greg-fer) [inline include/caf/leds.rst:61]:
```suggestion
If only one HW LED is used for a monochromatic setting, the module converts the tri-channel color to a single value of brightness and passes it to this single HW LED.
```

File: include/caf/leds.rst [hunk 6]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
 
+Configuring GPIO LEDs
+---------------------
+
+To properly use the LEDs module and have LEDs driven by GPIO, you must configure the GPIO driver and the LED GPIO driver.
+
+Enabling the GPIOs
+~~~~~~~~~~~~~~~~~~
+
+The Zephyr boards in general configure and enable the GPIO drivers by default and no additional configuration in needed.
```
Comment (greg-fer) [inline include/caf/leds.rst:183]:
```suggestion
In general, boards in Zephyr configure and enable the GPIO drivers by default, so no additional configuration is needed.
```

File: include/caf/leds.rst [hunk 7]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
+The LED GPIO driver can also use one of the GPIO expanders supported by Zephyr.
```
Comment (greg-fer) [inline include/caf/leds.rst:184]:
```suggestion
However, you can have the LED GPIO driver use one of the GPIO expanders supported by Zephyr.
```

File: include/caf/leds.rst [hunk 8]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
+For example, DTS configuration of ``thingy52_nrf52832`` board supports ``sx1509b`` GPIO expander that is used to control lightwell RGB LEDs.
```
Comment (greg-fer) [inline include/caf/leds.rst:185]:
```suggestion
For example, the DTS configuration of the ``thingy52_nrf52832`` board supports ``sx1509b`` GPIO expander, which is used to control lightwell RGB LEDs.
```

File: include/caf/leds.rst [hunk 9]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
+
+Enabling the LED GPIO nodes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The configuration process requires enabling or creating the LED GPIO nodes.
+The LED GPIO is configured as node that is compatible with ``gpio-leds``.
```
Comment (greg-fer) [inline include/caf/leds.rst:191]:
```suggestion
The LED GPIO is configured as a node that is compatible with ``gpio-leds``.
```

File: include/caf/leds.rst [hunk 10]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
+			gpios = <&sx1509b 5 GPIO_ACTIVE_LOW>;
+			label = "Green LED";
+		};
+		led2: led_2 {
+			gpios = <&sx1509b 6 GPIO_ACTIVE_LOW>;
+			label = "Blue LED";
+		};
+	};
+
+* Example 2 - Monochromatic LED connected directly to GPIO of MCU
```
Comment (greg-fer) [inline include/caf/leds.rst:217]:
```suggestion
* Example 2 - Monochromatic LED connected directly to the GPIO of an MCU
```

File: include/caf/leds.rst [hunk 11]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
+		led0: led_0 {
+			gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
+			label = "Green LED 0";
+		};
+	};
+
+Make sure to configure all the LED GPIO nodes that are used by the application.
+
+.. note::
+   Boards defined by Zephyr in general define a ``leds`` node that is compatible with ``gpio-leds``.
```
Comment (greg-fer) [inline include/caf/leds.rst:235]:
```suggestion
   In general, boards defined by Zephyr define a ``leds`` node that is compatible with ``gpio-leds``.
```

File: include/caf/leds.rst [hunk 12]
```diff
@@ -158,10 +172,74 @@ The following code snippets show examples of how the file contents could look in
...
+   The node can also be used by the module.
+   Before enabling the node, make sure to update it if it's needed to meet the requirements on CAF LEDs module.
```
Comment (greg-fer) [inline include/caf/leds.rst:237]:
```suggestion
   Before enabling the node, make sure it meets the requirements of the CAF LEDs module.
```

File: include/caf/leds.rst [hunk 13]
```diff
@@ -170,6 +248,9 @@ The LED color is achieved by setting the proper pulse widths for the PWM signals
 To achieve the desired LED effect, colors for the given LED are periodically updated using work (:c:struct:`k_work_delayable`).
 One work automatically updates the color of a single LED.
 
+.. note::
+   If GPIO based implementation is used, the signal's duty cycle may be either 0% or 100% - the LED may be either turned ON or OFF.
```
Comment (greg-fer) [inline include/caf/leds.rst:252]:
```suggestion
   If you use the GPIO-based implementation, the signal's duty cycle can be either 0% or 100% and the LED can be either turned on or off.
```

File: include/caf/leds.rst [hunk 14]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
 The following Kconfig options are also available for this module:
 
 * :option:`CONFIG_CAF_LEDS_PM_EVENTS` - This option enables the reaction to `Power management events`_.
 
-Configuring PWM ports and LEDs
-==============================
+.. note::
+   The LED GPIO driver implementation supports only turning LED ON or OFF.
+   Smooth changes of brightness are not supported, because of hardware limitations.
+   Therefore, the LED effects that use smooth changes of LED brightness are not properly displayed by this implementation.
```
Comment (greg-fer) [inline include/caf/leds.rst:43]: Wait, but in line 31 you claim more or less the opposite. Maybe we need to say it directly not to use GPIO if you're interested in brightness changes?

File: include/caf/leds.rst (hunk 2 above)
Comment (greg-fer) [reply include/caf/leds.rst:24]: Yes, this is a good point.

**Review by greg-fer — CHANGES_REQUESTED** [review]

File: include/caf/leds.rst (hunk 1 above)
Comment (MarekPieta) [reply include/caf/leds.rst:31]: There is a note with more detailed explanation of this limitation. The CAF module still uses API to set the LED brightness (but it actually can set the brightness either to 100% or 0% if GPIO LED driver is in use).

File: include/caf/leds.rst (hunk 2 above)
Comment (MarekPieta) [reply include/caf/leds.rst:24]: Changed

File: include/caf/leds.rst [hunk 15]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
 The LEDs module of the :ref:`lib_caf` (CAF) is responsible for controlling LEDs in response to LED effect set by a ``led_event``.
 The source of such events could be any other module in |NCS|.
 
-The module uses its own variant of Zephyr's :ref:`zephyr:pwm_api` driver for setting the LED color, either RGB or monochromatic.
-For this reason, it requires a manual configuration of the PWM ports and LEDs.
+The module uses Zephyr's :ref:`zephyr:led_api` driver for setting the LED color, either RGB or monochromatic.
+The Zephyr's LED driver can use either GPIO or PWM based implementation.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply include/caf/leds.rst:14]: Ok I will add a short explanation with example usage here.

File: include/caf/leds.rst (hunk 14 above)
Comment (MarekPieta) [reply include/caf/leds.rst:43]: Turning LED on and off is also actually done by setting the brightness to a selected level

File: include/caf/leds.rst [hunk 16]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
-==============================
+.. note::
+   The LED GPIO driver implementation supports only turning LED ON or OFF.
+   Smooth changes of brightness are not supported, because of hardware limitations.
+   Therefore, the LED effects that use smooth changes of LED brightness are not properly displayed by this implementation.
+
+Configuring LEDs in DTS
+=======================
+
+You can configure LEDs using compatible DTS nodes.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply include/caf/leds.rst:48]: Ok I will rephrase this one here and remove sentences that appear before.

File: include/caf/leds.rst [hunk 17]
```diff
@@ -10,27 +10,59 @@ CAF: LEDs module
...
+The configuration process requires enabling or creating DTS nodes.
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply include/caf/leds.rst:49]: LEDs in Zephyr (both GPIO-based and PWM-based) are configured using compatible DTS nodes. I will add this information too.

**Review by greg-fer — APPROVED** [review]

**Review by emob-nordic — APPROVED** [review]

**Review by pdunaj — APPROVED** [review]

**Comment by MarekPieta** [comment]
Emil asked me to use list in configuration. The change was applied

**Pull request merged**